src/providers/hana/qgshanaprovider.cpp

```diff
@@ -0,0 +1,1780 @@
+}
+
+bool QgsHanaProvider::addAttributes( const QList<QgsField> &attributes )
+{
+  if ( attributes.isEmpty() )
+    return true;
+
+  QgsHanaConnectionRef conn( mUri );
+  if ( conn.isNull() )
+    return false;
+
+  QString columnDefs;
+  for ( const QgsField &field : attributes )
+  {
+    if ( !columnDefs.isEmpty() )
+      columnDefs += QStringLiteral( "," );
+
+    columnDefs += QgsHanaUtils::quotedIdentifier( field.name() ) + " " + field.typeName();
+
+    if ( !field.comment().isEmpty() )
+      columnDefs += QStringLiteral( " COMMENT " ) + QgsHanaUtils::quotedString( field.comment() );
+  }
+
+  QString sql = QStringLiteral( "ALTER TABLE %1.%2 ADD (%3)" ).arg(
+                  QgsHanaUtils::quotedIdentifier( mSchemaName ), QgsHanaUtils::quotedIdentifier( mTableName ), columnDefs );
+
+  try
+  {
+    conn->execute( sql );
+    conn->commit();
+  }
+  catch ( const QgsHanaException &ex )
+  {
+    pushError( tr( "HANA failed to add feature: %1" )
+               .arg( QgsHanaUtils::formatErrorMessage( ex.what(), false ) ) );
+    conn->rollback();
+    return false;
+  }
+
+  readAttributeFields();
+
+  return true;
+}
+
+bool QgsHanaProvider::deleteAttributes( const QgsAttributeIds &attributes )
+{
+  if ( attributes.isEmpty() )
+    return false;
+
+  QgsHanaConnectionRef conn( mUri );
+  if ( conn.isNull() )
+    return false;
+
+  QString columnNames;
+  for ( int attrId : attributes )
+  {
+    if ( !columnNames.isEmpty() )
+      columnNames += QStringLiteral( "," );
+    const QgsField &field = mAttributeFields.at( attrId );
+    columnNames += QStringLiteral( "%1" ).arg( QgsHanaUtils::quotedIdentifier( field.name() ) );
+  }
+
+  QString sql = QStringLiteral( "ALTER TABLE %1.%2 DROP (%3)" ).arg(
+                  QgsHanaUtils::quotedIdentifier( mSchemaName ), QgsHanaUtils::quotedIdentifier( mTableName ), columnNames );
+
+  try
+  {
+    conn->execute( sql );
+    conn->commit();
+  }
+  catch ( const QgsHanaException &ex )
+  {
+    pushError( tr( "HANA error while deleting attributes: %1" )
+               .arg( QgsHanaUtils::formatErrorMessage( ex.what(), false ) ) );
+    conn->rollback();
+    return false;
+  }
+
+  readAttributeFields();
+
+  return true;
+}
+
+bool QgsHanaProvider::renameAttributes( const QgsFieldNameMap &fieldMap )
+{
+  if ( mIsQuery )
+    return false;
+
+  QgsHanaConnectionRef conn( mUri );
+  if ( conn.isNull() )
+    return false;
+
+  try
+  {
+    for ( QgsFieldNameMap::const_iterator it = fieldMap.begin(); it != fieldMap.end(); ++it )
+    {
+      int fieldIndex = it.key();
+      if ( fieldIndex < 0 || fieldIndex >= mAttributeFields.count() )
+      {
+        pushError( tr( "Invalid attribute index: %1" ).arg( fieldIndex ) );
+        return false;
+      }
+
+      if ( mAttributeFields.indexFromName( it.value() ) >= 0 )
+      {
+        pushError( tr( "Error renaming field %1: name '%2' already exists" ).arg( fieldIndex ).arg( it.value() ) );
+        return false;
+      }
```

Comment:
I guess this check does not always do what it is supposed to do as the metadata is not updated after each column update.

Imagine we have a table with columns `A` and `B`. Now, renaming `B` to `C` and then renaming `A` to `B` (in one call to `renameAttributes`) would succeed, but this check prevents it from succeeding as it still sees column `B` after it has been renamed to `C`.

On the other hand, if the request was to rename `A` to `C` and `B` to `C`, this check would not be able to detect the conflict.

Maybe we should just remove it. HANA will report an error then. Downside of that is that DDL statements are not undone by calling `rollback()`. Hence, we could end-up with a partial renaming. But that's not so bad. No data is lost and the renaming can be undone manually by the user.

Reply:
Reworked in 599ba2735.